test/teste2efiletransfer.cpp

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) by Claudio Cambra <claudio.cambra@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QObject>
+#include <QTest>
+#include <QSignalSpy>
+
+#include "gui/accountstate.h"
+#include "gui/folderman.h"
+#include "common/utility.h"
+
+#include "endtoendtestutils.h"
+
+class E2eFileTransferTest : public QObject
+{
+    Q_OBJECT
+
+public:
+    E2eFileTransferTest() = default;
+
+private:
+    EndToEndTestHelper _helper;
+    OCC::Folder *_testFolder;
+
+private slots:
+    void initTestCase()
+    {
+        QSignalSpy accountReady(&_helper, &EndToEndTestHelper::accountReady);
+        _helper.startAccountConfig();
+        QVERIFY(accountReady.wait(3000));
+
+        const auto accountState = _helper.accountState();
+        QSignalSpy accountConnected(accountState.data(), &OCC::AccountState::isConnectedChanged);
+        QVERIFY(accountConnected.wait(30000));
+
+        _testFolder = _helper.configureSyncFolder();
+        QVERIFY(_testFolder);
+    }
+
+    void testSyncFolder()
+    {
+        // Try the down-sync first
+        QSignalSpy folderSyncFinished(_testFolder, &OCC::Folder::syncFinished);
+        OCC::FolderMan::instance()->forceSyncForFolder(_testFolder);
+        QVERIFY(folderSyncFinished.wait(3000));
+
+        const auto testFolderPath = _testFolder->path();
+        const QString expectedFilePath(testFolderPath + QStringLiteral("welcome.txt"));
+        const QFile expectedFile(expectedFilePath);
+        qDebug() << "Checking if expected file exists at:" << expectedFilePath;
+        QVERIFY(expectedFile.exists());
+
+        // Now write a file to test the upload
+        const auto fileName = QStringLiteral("test_file.txt");
+        const QString localFilePath(_testFolder->path() + fileName);
+        QVERIFY(OCC::Utility::writeRandomFile(localFilePath));
+
+        OCC::FolderMan::instance()->forceSyncForFolder(_testFolder);
+        QVERIFY(folderSyncFinished.wait(3000));
+        qDebug() << "First folder sync complete";
+
+        const auto waitForServerToProcessTime = QTime::currentTime().addSecs(3);
+        while (QTime::currentTime() < waitForServerToProcessTime) {
+            QCoreApplication::processEvents(QEventLoop::AllEvents, 100);
+        }
```

Comment:
why is this needed ?

Reply:
The server will often not be able to process new files immediately after the upload/delete is done, so we sleep for three seconds to give the server a chance to finish what it's doing with the recent client's action, and then we check the action we wanted was completed

Reply:
do not we take the risk of random failures ?
how does the files client for android handle this ?

Reply:
Android tests also make use of waiting through "sleep":

https://github.com/nextcloud/android/blob/90306642627e878789570a96bc415f0dd18f5d2e/app/src/androidTest/java/com/owncloud/android/FileIT.java#L117

https://github.com/nextcloud/android/blob/90306642627e878789570a96bc415f0dd18f5d2e/app/src/androidTest/java/com/owncloud/android/files/services/FileUploaderIT.kt#L151